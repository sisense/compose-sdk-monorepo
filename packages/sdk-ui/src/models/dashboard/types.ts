import { CommonFiltersOptions } from '@/common-filters/types';
import { ColorPaletteTheme, TabberConfig } from '@/types';

export type {
  CommonFiltersOptions,
  CommonFiltersApplyMode,
  FiltersIgnoringRules,
} from '@/common-filters/types';

/**
 * Column layout of dashboard widgets
 */
export interface WidgetsPanelColumnLayout {
  columns: {
    widthPercentage: number;
    rows: {
      cells: {
        widthPercentage: number;
        /**
         * @privateRemarks
         * This value appears to be overwritten by the widget's height property.
         * Marking as optional and internal for now.
         * @internal
         */
        height?: number | string;
        widgetId: string;
      }[];
    }[];
  }[];
}

/**
 * Layout of dashboard widgets panel, which is a union of different layout algorithms
 */
export type WidgetsPanelLayout = WidgetsPanelColumnLayout;

/**
 * Widget ID
 */
export type WidgetId = string;
/**
 * Options for widgets in a dashboard
 *
 * For example, how common filters defined at the dashboard level should be applied to widgets.
 */
export type WidgetsOptions = Record<WidgetId, { filtersOptions?: CommonFiltersOptions }>;
/**
 * Options for TabberWidetss in a dashboard
 *
 * This property actually moves responsibility on the layout management from the tabber widgets to the dashboard,
 * storing all the tabbers configs in the single place
 * @internal
 */
export type TabbersOptions = Record<WidgetId, TabberConfig>;

/**
 * Style options for the dashboard.
 */
export type DashboardStyleOptions = {
  /** Collection of colors used to color various elements */
  palette?: ColorPaletteTheme;
  /** Background color */
  backgroundColor?: string;
  /** Width of the divider line between widgets */
  dividerLineWidth?: number;
  /** Divider line color */
  dividerLineColor?: string;
};
