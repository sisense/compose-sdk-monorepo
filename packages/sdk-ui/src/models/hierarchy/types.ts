import type { Attribute, DataSource } from '@sisense/sdk-data';

/**
 * Options for retrieving hierarchies.
 *
 * @internal
 */
export type GetHierarchiesOptions = {
  /**
   * The dimension for which to retrieve the hierarchies.
   */
  dimension: Attribute;
  /**
   * The data source from which to retrieve the hierarchies - e.g. `Sample ECommerce`.
   *
   * If not specified, the query will use the `defaultDataSource` specified in the parent Sisense Context.
   */
  dataSource?: DataSource;
  /**
   * A list of hierarchy IDs to retrieve specific hierarchies.
   */
  ids?: string[];
  /**
   * A flag indicating whether to filter the retrieved hierarchies based on the `alwaysIncluded` field.
   *
   * When set to true, only hierarchies with `alwaysIncluded: true` will be returned.
   * When set to false, only hierarchies with `alwaysIncluded: false` will be returned.
   * If not specified, all hierarchies will be returned.
   */
  alwaysIncluded?: boolean;
};
