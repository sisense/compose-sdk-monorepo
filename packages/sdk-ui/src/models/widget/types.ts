import { ChartType } from '@/index';

/**
 * Widget type
 *
 * todo: as a future refactoring need to replace this type by combination of WidgetType and ChartType
 *
 * @internal
 */
export type WidgetTypeInternal = ChartType | 'pivot' | 'plugin' | 'text';
