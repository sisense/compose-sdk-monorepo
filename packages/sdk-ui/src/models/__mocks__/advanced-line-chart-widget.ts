import { WidgetDto } from '@/widget-by-id/types';

/**
 * Line chart widget DTO with advanced analytics (e.g. trend, forecast).
 */
export const advancedLineChartWidgetDto = {
  title: 'Total of Revenue - total of Cost by months in Date',
  type: 'chart/line',
  subtype: 'line/spline',
  desc: '',
  datasource: {
    address: 'LocalHost',
    title: 'Sample ECommerce',
    id: 'localhost_aSampleIAAaECommerce',
    live: false,
    lastBuildTime: '2025-03-27T14:17:42.257Z',
  },
  style: {
    lineWidth: {
      width: 'bold',
    },
    legend: {
      enabled: true,
      position: 'bottom',
    },
    seriesLabels: {
      enabled: false,
      rotation: 0,
    },
    markers: {
      enabled: false,
      fill: 'filled',
      size: 'small',
    },
    xAxis: {
      enabled: true,
      ticks: true,
      labels: {
        enabled: true,
        rotation: 0,
      },
      title: {
        enabled: false,
      },
      x2Title: {
        enabled: false,
      },
      gridLines: true,
      isIntervalEnabled: false,
    },
    yAxis: {
      inactive: false,
      enabled: true,
      ticks: true,
      labels: {
        enabled: true,
        rotation: 0,
      },
      title: {
        enabled: false,
      },
      gridLines: true,
      logarithmic: false,
      isIntervalEnabled: true,
    },
    y2Axis: {
      inactive: false,
      enabled: true,
      ticks: true,
      labels: {
        enabled: true,
        rotation: 0,
      },
      title: {
        enabled: false,
      },
      gridLines: false,
      logarithmic: false,
      isIntervalEnabled: true,
      templateMainYHasGridLines: true,
    },
    navigator: {
      enabled: true,
    },
    dataLimits: {
      seriesCapacity: 50,
      categoriesCapacity: 100000,
    },
    narration: {
      display: 'above',
      verbosity: 'low',
      labels: [
        {
          id: 'months_in_date',
          title: 'months in Date',
          singular: 'months in Date',
          plural: 'months in Date',
        },
      ],
    },
  },
  instanceid: '155A3-F652-14',
  realTimeRefreshing: false,
  metadata: {
    ignore: {
      dimensions: [],
      ids: [],
      all: false,
    },
    panels: [
      {
        name: 'x-axis',
        items: [
          {
            jaql: {
              title: 'months in Date',
              dim: '[Commerce.Date]',
              datatype: 'datetime',
              level: 'months',
            },
            instanceid: 'E1C7A-2647-40',
            panel: 'rows',
            format: {
              mask: {
                years: 'yyyy',
                quarters: 'yyyy Q',
                months: 'MM/yyyy',
                weeks: 'ww yyyy',
                days: 'shortDate',
                minutes: 'HH:mm',
                seconds: 'MM/dd/yyyy HH:mm:ss',
                dateAndTime: 'MM/dd/yyyy HH:mm',
                isdefault: true,
              },
            },
            hierarchies: ['calendar', 'calendar - weeks'],
          },
        ],
      },
      {
        name: 'values',
        items: [
          {
            jaql: {
              title: 'total of Revenue - total of Cost',
              formula: 'sum([12fa]) - sum([c1bb])',
              context: {
                '[12fa]': {
                  title: 'Revenue',
                  dim: '[Commerce.Revenue]',
                  datatype: 'numeric',
                },
                '[c1bb]': {
                  title: 'Cost',
                  dim: '[Commerce.Cost]',
                  datatype: 'numeric',
                },
              },
            },
            instanceid: 'D4056-8BBE-A1',
            panel: 'measures',
            format: {
              mask: {
                type: 'number',
                abbreviations: {
                  t: true,
                  b: true,
                  m: true,
                  k: true,
                },
                separated: true,
                decimals: 'auto',
                abbreviateAll: false,
                isdefault: true,
              },
              color: {
                type: 'color',
                color: '#d39a47',
                isHandPickedColor: true,
              },
            },
            statisticalModels: {
              trend: {
                isEnabled: true,
                isViewerDisabled: false,
                trendType: 'linear',
                ignoreAnomalies: false,
                trendOnForecast: false,
                compare: {
                  isEnabled: false,
                  period: 'year',
                },
                isAccessible: true,
              },
              forecast: {
                isEnabled: true,
                isViewerDisabled: false,
                explainVariable: null,
                evaluation: {
                  type: 'all',
                  numLastPointsForEvaluation: 0,
                  ignoreLast: 0,
                },
                forecastPeriod: 6,
                confidence: 80,
                modelType: 'en',
                boundaries: {
                  upper: {
                    isEnabled: false,
                    value: null,
                  },
                  lower: {
                    isEnabled: false,
                    value: null,
                  },
                  isInt: {
                    isEnabled: false,
                  },
                },
                isAccessible: true,
              },
            },
          },
          {
            jaql: {
              table: 'Commerce',
              column: 'Revenue',
              dim: '[Commerce.Revenue]',
              datatype: 'numeric',
              columnTitle: 'Revenue',
              tableTitle: 'Commerce',
              agg: 'sum',
              title: 'Total Revenue',
            },
            instanceid: 'DA977-EB29-8B',
            panel: 'measures',
            format: {
              mask: {
                type: 'number',
                abbreviations: {
                  t: true,
                  b: true,
                  m: true,
                  k: true,
                },
                separated: true,
                decimals: 'auto',
                abbreviateAll: false,
                isdefault: true,
              },
              color: {
                type: 'color',
                color: '#26a0a2',
                isHandPickedColor: true,
              },
            },
            statisticalModels: {
              forecast: {
                isEnabled: true,
                isViewerDisabled: false,
                explainVariable: null,
                evaluation: {
                  type: 'all',
                  numLastPointsForEvaluation: 0,
                  ignoreLast: 0,
                },
                forecastPeriod: 6,
                confidence: 80,
                modelType: 'en',
                boundaries: {
                  upper: {
                    isEnabled: false,
                    value: null,
                  },
                  lower: {
                    isEnabled: false,
                    value: null,
                  },
                  isInt: {
                    isEnabled: false,
                  },
                },
                isAccessible: true,
              },
              trend: {
                isEnabled: true,
                isViewerDisabled: false,
                trendType: 'linear',
                ignoreAnomalies: false,
                trendOnForecast: false,
                compare: {
                  isEnabled: false,
                  period: 'year',
                },
                isAccessible: true,
              },
            },
            y2: true,
          },
        ],
      },
      {
        name: 'break by',
        items: [],
      },
      {
        name: 'filters',
        items: [],
      },
    ],
  },
  options: {
    dashboardFiltersMode: 'select',
    selector: true,
    triggersDomready: true,
    autoUpdateOnEveryChange: true,
    drillToAnywhere: true,
    previousScrollerLocation: {
      min: null,
      max: null,
    },
  },
} as unknown as WidgetDto;
