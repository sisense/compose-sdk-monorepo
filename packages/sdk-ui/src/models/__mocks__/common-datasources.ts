export const commonDataSources = [
  {
    address: 'LocalHost',
    title: 'Sample ECommerce',
    id: 'localhost_aSampleIAAaECommerce',
    database: 'aSampleIAAaECommerce',
    fullname: 'localhost/Sample ECommerce',
    live: false,
  },
];
