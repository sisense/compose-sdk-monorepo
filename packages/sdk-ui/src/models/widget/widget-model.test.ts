/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { widgetModelTranslator } from '.';
import { WidgetDto, WidgetType } from '../../dashboard-widget/types';
import { TranslatableError } from '../../translation/translatable-error';
import { sampleEcommerceDashboard as dashboardMock } from '../__mocks__/sample-ecommerce-dashboard';
import { sampleHealthcareDashboard } from '../__mocks__/sample-healthcare-dashboard';
import { dashboardWithTextWidget } from '../__mocks__/dashboard-with-text-widget';

describe('WidgetModel', () => {
  let mockWidgetDto: WidgetDto;

  beforeEach(() => {
    // Create a mock WidgetDto object with necessary properties for testing
    mockWidgetDto = dashboardMock.widgets![0];
  });

  describe('constructor', () => {
    it('should create a WidgetModel instance with valid input', () => {
      const widget = widgetModelTranslator.fromWidgetDto(mockWidgetDto);

      expect(widget.oid).toBe(mockWidgetDto.oid);
      expect(widget.title).toBe(mockWidgetDto.title);
      expect(widget.description).toBe(mockWidgetDto.desc || '');
      expect(widget.dataSource).toStrictEqual({
        title: mockWidgetDto.datasource.fullname || mockWidgetDto.datasource.title,
        type: 'elasticube',
      });
      expect(widget.widgetType).toBe(mockWidgetDto.type);
      expect(widget.dataOptions).toMatchObject({
        max: expect.arrayContaining([expect.objectContaining({ column: expect.any(Object) })]),
        min: expect.arrayContaining([expect.objectContaining({ column: expect.any(Object) })]),
        value: expect.arrayContaining([expect.objectContaining({ column: expect.any(Object) })]),
      });
      expect(widget.styleOptions).toMatchObject({
        subtype: 'indicator/gauge',
        skin: 1,
        indicatorComponents: {
          labels: { shouldBeShown: true },
          ticks: { shouldBeShown: true },
          title: { shouldBeShown: false, text: expect.any(String) },
          secondaryTitle: { text: '' },
        },
      });
      expect(widget.filters).toHaveLength(0);
      expect(widget.drilldownOptions).toMatchObject({
        drilldownPaths: expect.arrayContaining([]),
        drilldownSelections: expect.arrayContaining([]),
      });
    });

    it('should create plugin widget when unsupported widget type', () => {
      const unsupportedWidgetDto: WidgetDto = {
        ...mockWidgetDto,
        type: 'unsupportedType' as WidgetType,
      };

      const widgetPlugin = widgetModelTranslator.fromWidgetDto(unsupportedWidgetDto);

      expect(widgetPlugin.widgetType).equals('plugin');
      expect(widgetPlugin.pluginType).equals(unsupportedWidgetDto.type);
      expect(widgetPlugin.dataOptions).toBeDefined();
      expect(widgetPlugin.styleOptions).equals(unsupportedWidgetDto.style);
    });
  });

  describe('getExecuteQueryParams', () => {
    it('returns execute query parameters for chart widget', () => {
      const widget = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
      const executeQueryParams = widget.getExecuteQueryParams();

      expect(executeQueryParams).toMatchObject({
        dataSource: { title: 'Sample ECommerce', type: 'elasticube' },
        dimensions: [],
        measures: expect.arrayContaining([
          expect.objectContaining({
            name: expect.any(String),
            type: 'basemeasure',
          }),
          expect.objectContaining({
            name: expect.any(String),
            type: 'calculatedmeasure',
          }),
          expect.objectContaining({
            name: expect.any(String),
            type: 'calculatedmeasure',
          }),
        ]),
      });
    });

    it('returns execute query parameters for table widget', () => {
      const tableWidgetModel = widgetModelTranslator.fromWidgetDto(
        sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'tablewidget')!,
      );
      const executeQueryParams = tableWidgetModel.getExecuteQueryParams();

      expect(executeQueryParams).toMatchObject({
        dataSource: { title: 'Sample Healthcare', type: 'elasticube' },
        dimensions: [expect.objectContaining({ name: 'DIAGNOSIS' })],
        measures: expect.arrayContaining([
          expect.objectContaining({
            name: '# PATIENTS',
            type: 'basemeasure',
          }),
          expect.objectContaining({
            name: 'AVG COST',
            type: 'basemeasure',
          }),
        ]),
        filters: [
          {
            attribute: {
              id: '[Diagnosis.Description]',
            },
            type: 'filter',
          },
        ],
      });
    });

    it('should throw an error for pivot widget', () => {
      const pivotWidgetModel = widgetModelTranslator.fromWidgetDto(
        sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'pivot')!,
      );

      expect(() => pivotWidgetModel.getExecuteQueryParams()).toThrow(TranslatableError);
    });
  });

  describe('getExecutePivotQueryParams', () => {
    it('returns execute query parameters for pivot widget', () => {
      const pivotWidgetModel = widgetModelTranslator.fromWidgetDto(
        sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'pivot')!,
      );
      const executePivotQueryParams = pivotWidgetModel.getExecutePivotQueryParams();

      expect(executePivotQueryParams).toMatchObject({
        dataSource: { title: 'Sample Healthcare', type: 'elasticube' },
        rows: [{ attribute: expect.objectContaining({ name: 'DIAGNOSIS' }) }],
        values: expect.arrayContaining([
          expect.objectContaining({
            measure: expect.objectContaining({
              name: '# PATIENTS',
              type: 'basemeasure',
            }),
          }),
          expect.objectContaining({
            measure: expect.objectContaining({
              name: 'AVG COST',
              type: 'basemeasure',
            }),
          }),
          expect.objectContaining({
            measure: expect.objectContaining({
              name: 'AVG DAYS ADMITTED',
              type: 'calculatedmeasure',
            }),
          }),
        ]),
        filters: [
          {
            attribute: {
              id: '[Diagnosis.Description]',
            },
            type: 'filter',
          },
        ],
      });
    });

    it('should throw an error for non-pivot widget', () => {
      const nonPivotWidgetModel = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
      expect(() => nonPivotWidgetModel.getExecutePivotQueryParams()).toThrow(TranslatableError);
    });

    describe('getChartProps', () => {
      it('should throw an error for pivot widget', () => {
        const pivotWidgetModel = widgetModelTranslator.fromWidgetDto(
          sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'pivot')!,
        );
        expect(() => pivotWidgetModel.getChartProps()).toThrow(TranslatableError);
      });
      it('should throw an error for text widget', () => {
        const textWidgetModel = widgetModelTranslator.fromWidgetDto(
          dashboardWithTextWidget.widgets!.find((widget) => widget.type === 'richtexteditor')!,
        );
        expect(() => textWidgetModel.getChartProps()).toThrow(TranslatableError);
      });
      it('should return chart props for chart widgets', () => {
        const widget = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
        const chartProps = widget.getChartProps();

        expect(chartProps).toMatchObject({
          chartType: 'indicator',
          dataOptions: expect.objectContaining({
            max: expect.arrayContaining([expect.objectContaining({ column: expect.any(Object) })]),
            min: expect.arrayContaining([expect.objectContaining({ column: expect.any(Object) })]),
            value: expect.arrayContaining([
              expect.objectContaining({ column: expect.any(Object) }),
            ]),
          }),
          styleOptions: expect.objectContaining({
            skin: 1,
            subtype: 'indicator/gauge',
            indicatorComponents: expect.objectContaining({
              labels: expect.objectContaining({
                shouldBeShown: true,
              }),
              ticks: expect.objectContaining({
                shouldBeShown: true,
              }),
              title: expect.objectContaining({
                shouldBeShown: false,
                text: expect.any(String),
              }),
              secondaryTitle: expect.objectContaining({
                text: '',
              }),
            }),
          }),
          dataSet: { title: 'Sample ECommerce', type: 'elasticube' },
          filters: [],
        });
      });

      it('should return chart props for table widget', () => {
        const tableWidgetDto = sampleHealthcareDashboard.widgets!.find(
          (widget) => widget.type === 'tablewidget',
        )!;
        const tableWidget = widgetModelTranslator.fromWidgetDto(tableWidgetDto);
        const chartProps = tableWidget.getChartProps();

        expect(chartProps).toMatchSnapshot();
      });
    });

    describe('getTableProps', () => {
      it('should throw an error for non-table widgets', () => {
        const nonTableWidgetModel = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
        expect(() => nonTableWidgetModel.getTableProps()).toThrow(TranslatableError);
      });
      it('should return TableProps for table widgets', () => {
        const tableWidgetModel = widgetModelTranslator.fromWidgetDto(
          sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'tablewidget')!,
        );
        const tableProps = tableWidgetModel.getTableProps();

        expect(tableProps).toMatchObject({
          dataOptions: expect.objectContaining({
            columns: expect.arrayContaining([
              expect.objectContaining({
                column: expect.objectContaining({
                  name: 'DIAGNOSIS',
                }),
              }),
              expect.objectContaining({
                column: expect.objectContaining({
                  name: '# PATIENTS',
                }),
              }),
              expect.objectContaining({
                column: expect.objectContaining({
                  name: 'AVG COST',
                }),
              }),
            ]),
          }),
          styleOptions: expect.objectContaining({
            columns: { alternatingColor: { enabled: expect.any(Boolean) } },
            rows: { alternatingColor: { enabled: expect.any(Boolean) } },
            header: { color: { enabled: expect.any(Boolean) } },
          }),
          dataSet: { title: 'Sample Healthcare', type: 'elasticube' },
          filters: [
            {
              attribute: {
                id: '[Diagnosis.Description]',
              },
            },
          ],
        });
      });
    });

    describe('getPivotTableProps', () => {
      it('should throw an error for non-pivot widgets', () => {
        const nonPivotWidgetModel = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
        expect(() => nonPivotWidgetModel.getPivotTableProps()).toThrow(TranslatableError);
      });
      it('should return PivotTableProps for pivot widgets', () => {
        const pivotWidgetModel = widgetModelTranslator.fromWidgetDto(
          sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'pivot')!,
        );
        const pivotProps = pivotWidgetModel.getPivotTableProps();

        expect(pivotProps).toMatchObject({
          dataOptions: expect.objectContaining({
            rows: [
              expect.objectContaining({ column: expect.objectContaining({ name: 'DIAGNOSIS' }) }),
            ],
            values: expect.arrayContaining([
              expect.objectContaining({
                column: expect.objectContaining({
                  name: '# PATIENTS',
                }),
              }),
              expect.objectContaining({
                column: expect.objectContaining({
                  name: 'AVG COST',
                }),
              }),
              expect.objectContaining({
                column: expect.objectContaining({
                  name: 'AVG DAYS ADMITTED',
                }),
              }),
            ]),
          }),
          styleOptions: expect.objectContaining({
            alternatingColumnsColor: expect.any(Boolean),
            alternatingRowsColor: expect.any(Boolean),
            headersColor: expect.any(Boolean),
          }),
          dataSet: { title: 'Sample Healthcare', type: 'elasticube' },
          filters: [
            {
              attribute: {
                id: '[Diagnosis.Description]',
              },
              type: 'filter',
            },
          ],
        });
      });
    });

    describe('getChartWidgetProps', () => {
      it('should return chart widget props correctly', () => {
        const widget = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
        const chartWidgetProps = widget.getChartWidgetProps();

        expect(chartWidgetProps).toMatchObject({
          // validate matching only basic shape of the object
          chartType: 'indicator',
          dataOptions: expect.objectContaining({
            value: expect.arrayContaining([
              expect.objectContaining({
                column: expect.objectContaining({
                  name: '# of unique Brand ID',
                  type: 'basemeasure',
                }),
              }),
            ]),
            secondary: [],
            min: expect.arrayContaining([
              expect.objectContaining({
                column: expect.objectContaining({
                  type: 'calculatedmeasure',
                }),
              }),
            ]),
            max: expect.arrayContaining([
              expect.objectContaining({
                column: expect.objectContaining({
                  type: 'calculatedmeasure',
                }),
              }),
            ]),
          }),
          styleOptions: expect.objectContaining({
            subtype: 'indicator/gauge',
            skin: 1,
            indicatorComponents: expect.any(Object),
          }),
          dataSource: { title: 'Sample ECommerce', type: 'elasticube' },
          filters: [],
          title: 'TOTAL BRANDS',
          description: '',
          drilldownOptions: {
            drilldownPaths: [],
            drilldownSelections: [],
          },
        });
      });

      it('should throw an error for non-chart widgets', () => {
        const tableWidgetModel = widgetModelTranslator.fromWidgetDto(
          sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'pivot')!,
        );

        expect(() => {
          tableWidgetModel.getChartWidgetProps();
        }).toThrow(TranslatableError);

        const textWidgetModel = widgetModelTranslator.fromWidgetDto(
          dashboardWithTextWidget.widgets!.find((widget) => widget.type === 'richtexteditor')!,
        );

        expect(() => {
          textWidgetModel.getChartWidgetProps();
        }).toThrow(TranslatableError);
      });
    });
  });

  describe('getTableWidgetProps', () => {
    it('should return table widget props correctly for table widget', () => {
      const tableWidgetModel = widgetModelTranslator.fromWidgetDto(
        sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'tablewidget')!,
      );
      const tableWidgetProps = tableWidgetModel.getTableWidgetProps();

      expect(tableWidgetProps).toMatchObject({
        dataOptions: expect.objectContaining({
          columns: expect.arrayContaining([
            expect.objectContaining({
              column: expect.objectContaining({
                name: 'DIAGNOSIS',
              }),
            }),
            expect.objectContaining({
              column: expect.objectContaining({
                name: '# PATIENTS',
              }),
            }),
            expect.objectContaining({
              column: expect.objectContaining({
                name: 'AVG COST',
              }),
            }),
          ]),
        }),
        styleOptions: expect.objectContaining({
          columns: { alternatingColor: { enabled: expect.any(Boolean) } },
          rows: { alternatingColor: { enabled: expect.any(Boolean) } },
          header: { color: { enabled: expect.any(Boolean) } },
        }),
        dataSource: { title: 'Sample Healthcare', type: 'elasticube' },
        filters: [
          {
            attribute: {
              id: '[Diagnosis.Description]',
            },
            type: 'filter',
          },
        ],
        title: 'TOP 10 DIAGNOSIS (table widget)',
        description: '',
      });
    });
    it('should throw an error for non-table widgets', () => {
      const nonTableWidgetModel = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
      expect(() => nonTableWidgetModel.getTableWidgetProps()).toThrow(TranslatableError);

      const textWidgetModel = widgetModelTranslator.fromWidgetDto(
        dashboardWithTextWidget.widgets!.find((widget) => widget.type === 'richtexteditor')!,
      );
      expect(() => textWidgetModel.getTableWidgetProps()).toThrow(TranslatableError);
    });
  });

  describe('getPivotTableWidgetProps', () => {
    it('should throw an error for non-pivot widgets', () => {
      const nonPivotWidgetModel = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
      expect(() => nonPivotWidgetModel.getPivotTableWidgetProps()).toThrow(TranslatableError);

      const textWidgetModel = widgetModelTranslator.fromWidgetDto(
        dashboardWithTextWidget.widgets!.find((widget) => widget.type === 'richtexteditor')!,
      );
      expect(() => textWidgetModel.getPivotTableWidgetProps()).toThrow(TranslatableError);
    });
    it('should return PivotTableWidgetProps for pivot widgets', () => {
      const pivotWidgetModel = widgetModelTranslator.fromWidgetDto(
        sampleHealthcareDashboard.widgets!.find((widget) => widget.type === 'pivot')!,
      );
      const pivotProps = pivotWidgetModel.getPivotTableWidgetProps();

      expect(pivotProps).toMatchObject({
        dataOptions: expect.objectContaining({
          rows: [
            expect.objectContaining({ column: expect.objectContaining({ name: 'DIAGNOSIS' }) }),
          ],
          values: expect.arrayContaining([
            expect.objectContaining({
              column: expect.objectContaining({
                name: '# PATIENTS',
              }),
            }),
            expect.objectContaining({
              column: expect.objectContaining({
                name: 'AVG COST',
              }),
            }),
            expect.objectContaining({
              column: expect.objectContaining({
                name: 'AVG DAYS ADMITTED',
              }),
            }),
          ]),
        }),
        styleOptions: expect.objectContaining({
          alternatingColumnsColor: expect.any(Boolean),
          alternatingRowsColor: expect.any(Boolean),
          headersColor: expect.any(Boolean),
        }),
        dataSource: { title: 'Sample Healthcare', type: 'elasticube' },
        filters: [
          {
            attribute: {
              id: '[Diagnosis.Description]',
            },
            type: 'filter',
          },
        ],
        title: 'TOP 10 DIAGNOSIS',
        description: '',
      });
    });
  });

  describe('getTextWidgetProps', () => {
    it('should return text widget props correctly for text widget', () => {
      const textWidgetModel = widgetModelTranslator.fromWidgetDto(
        dashboardWithTextWidget.widgets!.find((widget) => widget.type === 'richtexteditor')!,
      );
      const textWidgetProps = textWidgetModel.getTextWidgetProps();

      expect(textWidgetProps).toMatchObject({
        styleOptions: {
          bgColor: '#FFFFFF',
          html: '<font color="#008c9c" size="5"><u style="">fancy text</u></font>',
          textAlign: 'center',
          vAlign: 'valign-middle',
        },
      });
    });
    it('should throw an error for non-text widgets', () => {
      const nonTextWidgetModel = widgetModelTranslator.fromWidgetDto(mockWidgetDto);
      expect(() => nonTextWidgetModel.getTextWidgetProps()).toThrow(TranslatableError);
    });
  });
});
