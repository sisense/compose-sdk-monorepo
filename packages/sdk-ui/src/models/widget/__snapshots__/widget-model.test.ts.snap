// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WidgetModel > getExecutePivotQueryParams > getChartProps > should return chart props for table widget 1`] = `
{
  "chartType": "table",
  "dataOptions": {
    "columns": [
      {
        "column": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "address": "localHost",
            "database": "aSampleIAAaHealthcare",
            "fullname": "LocalHost/Sample Healthcare",
            "id": "localhost_aSampleIAAaHealthcare",
            "title": "Sample Healthcare",
          },
          "desc": "",
          "expression": "[Diagnosis.Description]",
          "name": "DIAGNOSIS",
          "type": "text-attribute",
        },
        "isColored": false,
        "sortType": "sortNone",
      },
      {
        "column": {
          "__serializable": "DimensionalElement",
          "aggregation": "countDistinct",
          "attribute": {
            "__serializable": "DimensionalElement",
            "dataSource": {
              "address": "localHost",
              "database": "aSampleIAAaHealthcare",
              "fullname": "LocalHost/Sample Healthcare",
              "id": "localhost_aSampleIAAaHealthcare",
              "title": "Sample Healthcare",
            },
            "desc": "",
            "expression": "[Admissions.Patient_ID]",
            "name": "# PATIENTS",
            "type": "numeric-attribute",
          },
          "dataSource": undefined,
          "desc": "",
          "name": "# PATIENTS",
          "sort": 0,
          "type": "basemeasure",
        },
        "numberFormatConfig": {
          "billion": false,
          "decimalScale": "auto",
          "kilo": false,
          "million": false,
          "name": "Numbers",
          "prefix": false,
          "symbol": undefined,
          "thousandSeparator": true,
          "trillion": false,
        },
        "sortType": "sortNone",
        "width": 134,
      },
      {
        "column": {
          "__serializable": "DimensionalElement",
          "aggregation": "avg",
          "attribute": {
            "__serializable": "DimensionalElement",
            "dataSource": {
              "address": "localHost",
              "database": "aSampleIAAaHealthcare",
              "fullname": "LocalHost/Sample Healthcare",
              "id": "localhost_aSampleIAAaHealthcare",
              "title": "Sample Healthcare",
            },
            "desc": "",
            "expression": "[Admissions.Cost_of_admission]",
            "name": "AVG COST",
            "type": "numeric-attribute",
          },
          "dataSource": undefined,
          "desc": "",
          "name": "AVG COST",
          "sort": 0,
          "type": "basemeasure",
        },
        "numberFormatConfig": {
          "billion": false,
          "decimalScale": 0,
          "kilo": false,
          "million": false,
          "name": "Currency",
          "prefix": true,
          "symbol": "$",
          "thousandSeparator": undefined,
          "trillion": false,
        },
        "sortType": "sortNone",
        "width": 182,
      },
    ],
  },
  "dataSet": {
    "address": "localHost",
    "id": "localhost_aSampleIAAaHealthcare",
    "title": "Sample Healthcare",
    "type": "elasticube",
  },
  "filters": [
    {
      "__serializable": "DimensionalElement",
      "attribute": {
        "__serializable": "DimensionalElement",
        "dataSource": {
          "address": "localHost",
          "database": "aSampleIAAaHealthcare",
          "fullname": "LocalHost/Sample Healthcare",
          "id": "localhost_aSampleIAAaHealthcare",
          "title": "Sample Healthcare",
        },
        "desc": "",
        "expression": "[Diagnosis.Description]",
        "name": "DESCRIPTION",
        "type": "text-attribute",
      },
      "count": 10,
      "dataSource": undefined,
      "desc": "",
      "filterType": "ranking",
      "measure": {
        "__serializable": "DimensionalElement",
        "aggregation": "countDistinct",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": undefined,
          "desc": "",
          "expression": "[Admissions.Patient_ID]",
          "name": "Patient_ID",
          "type": "numeric-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "name": "countDistinct Patient_ID",
        "sort": 0,
        "type": "basemeasure",
      },
      "name": "7a707cfbd72",
      "operator": "top",
      "type": "filter",
    },
  ],
  "styleOptions": {
    "columns": {
      "alternatingColor": {
        "enabled": false,
      },
    },
    "header": {
      "color": {
        "enabled": true,
      },
    },
    "rows": {
      "alternatingColor": {
        "enabled": true,
      },
    },
  },
}
`;
